src/topology/algebra/order/compact.lean

```diff
@@ -388,3 +389,29 @@ begin
 end
 
 end continuous_on
+
+lemma is_compact.exists_local_min_on_mem_subset {f : β → α} {s t : set β} {m : α} {z : β}
+  (ht : is_compact t) (hst : s ⊆ t) (hf : continuous_on f t) (hf1 : ∀ z ∈ t \ s, m ≤ f z)
+  (hz : z ∈ t) (hfz : f z < m) :
+  ∃ x ∈ s, is_local_min_on f t x :=
```

Comment:
Is this still true if you replace `(hf1 : ∀ z ∈ t \ s, m ≤ f z) (hz : z ∈ t) (hfz : f z < m)` by `(hz : z ∈ t) (hfz : ∀ z' ∈ t \ s, f z < f z')`? It's a strictly weaker assumption so I could be wrong, but looking at the proof I don't see where it would fail.

Also, could you add a comment/docstring explaining your use case? I know this is a classical argument to use in combination with the local max/min principle, but if you don't have it in mind this might look quite random.

(Same remarks for the lemma below)

Reply:
It will certainly work if `t \ s` is compact. But what if `f` reaches its minimum exactly once on the sphere and that point happens to be in `s` while the rest is in `t \ s`? OTOH assuming that `s` is open is the typical use case anyway.

Reply:
I don't see why that gives a counterexample, sorry :sweat_smile: Won't the minimum of `f` be the local minimum on `t` belonging to `s` that we are searching for?

Reply:
Oh I see that's because your proof was actually a bit too convoluted, so it didn't work with the weakened assumption. This works:
```lean
lemma is_compact.exists_local_min_on_mem_subset {f : β → α} {s t : set β} {z : β}
  (ht : is_compact t) (hst : s ⊆ t) (hf : continuous_on f t) (hz : z ∈ t)
  (hfz : ∀ z' ∈ t \ s, f z < f z') :
  ∃ x ∈ s, is_local_min_on f t x :=
begin
  obtain ⟨x, hx, hfx⟩ : ∃ x ∈ t, ∀ y ∈ t, f x ≤ f y := ht.exists_forall_le ⟨z, hz⟩ hf,
  have key : ∀ ⦃y⦄, y ∈ t → (∀ z' ∈ t \ s, f y < f z') → y ∈ s := λ y hy hfy,
    by { by_contra; simpa using ((hfy y ((mem_diff y).mpr ⟨hy,h⟩))) },
  have h1 : ∀ z' ∈ t \ s, f x < f z' := λ z' hz', (hfx z hz).trans_lt (hfz z' hz'),
  have h2 : x ∈ s := key hx h1,
  refine ⟨x, h2, eventually_nhds_within_of_forall hfx⟩
end

lemma is_compact.exists_local_min_mem_open {f : β → α} {s t : set β} {z : β} (ht : is_compact t)
  (hst : s ⊆ t) (hf : continuous_on f t) (hz : z ∈ t) (hfz : ∀ z' ∈ t \ s, f z < f z')
  (hs : is_open s) :
  ∃ x ∈ s, is_local_min f x :=
begin
  obtain ⟨x, hx, hfx⟩ := ht.exists_local_min_on_mem_subset hst hf hz hfz,
  exact ⟨x, hx, hfx.is_local_min (filter.mem_of_superset (hs.mem_nhds hx) hst)⟩
end
```

Reply:
Good catch! Yes in my head the first step was always to take the min on `t \ s` and show that it was above `m` but indeed this is useless.